command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,813 @@
+			return
+		case err := <-renewer.DoneCh():
+			// This case covers renewal completion and renewal errors
+			if err != nil {
+				c.logger.Error("failed to renew secret", "error", err)
+				return
+			}
+			c.logger.Debug("renewal halted; evicting from cache", "path", req.Request.URL.Path)
+			return
+		case renewal := <-renewer.RenewCh():
+			// This case captures secret renewals. Renewed secret is updated in
+			// the cached index.
+			c.logger.Debug("renewal received; updating cache", "path", req.Request.URL.Path)
+			err = c.updateResponse(ctx, renewal)
+			if err != nil {
+				c.logger.Error("failed to handle renewal", "error", err)
+				return
+			}
+		case <-index.RenewCtxInfo.DoneCh:
+			// This case indicates the renewal process to shutdown and evict
+			// the cache entry. This is triggered when a specific secret
+			// renewal needs to be killed without affecting any of the derived
+			// context renewals.
+			c.logger.Debug("done channel closed")
+			return
+		}
+	}
+}
+
+func (c *LeaseCache) updateResponse(ctx context.Context, renewal *api.RenewOutput) error {
+	id := ctx.Value(contextIndexID).(string)
+
+	// Get the cached index using the id in the context
+	index, err := c.db.Get(cachememdb.IndexNameID, id)
+	if err != nil {
+		return err
+	}
+	if index == nil {
+		return fmt.Errorf("missing cache entry for id: %q", id)
+	}
+
+	// Read the response from the index
+	resp, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(index.Response)), nil)
+	if err != nil {
+		c.logger.Error("failed to deserialize response", "error", err)
+		return err
+	}
+
+	// Update the body in the reponse by the renewed secret
+	bodyBytes, err := json.Marshal(renewal.Secret)
+	if err != nil {
+		return err
+	}
+	if resp.Body != nil {
+		resp.Body.Close()
+	}
+	resp.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+	resp.ContentLength = int64(len(bodyBytes))
+
+	// Serialize the response
+	var respBytes bytes.Buffer
+	err = resp.Write(&respBytes)
+	if err != nil {
+		c.logger.Error("failed to serialize updated response", "error", err)
+		return err
+	}
+
+	// Update the response in the index and set it in the cache
+	index.Response = respBytes.Bytes()
+	err = c.db.Set(index)
+	if err != nil {
+		c.logger.Error("failed to cache the proxied response", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// computeIndexID results in a value that uniquely identifies a request
+// received by the agent. It does so by SHA256 hashing the serialized request
+// object containing the request path, query parameters and body parameters.
+func computeIndexID(req *SendRequest) (string, error) {
+	var b bytes.Buffer
+
+	// Serialze the request
+	if err := req.Request.Write(&b); err != nil {
+		return "", fmt.Errorf("failed to serialize request: %v", err)
+	}
+
+	// Reset the request body after it has been closed by Write
+	if req.Request.Body != nil {
+		req.Request.Body.Close()
+	}
+	req.Request.Body = ioutil.NopCloser(bytes.NewBuffer(req.RequestBody))
+
```

Comment:
Won't this function not compute an equivalent index if the parameters are given in a different order?

Reply:
Yes, we settled on being acceptable for this first iteration (I forget if it was over Slack or one of the sync-ups). The expectation is that clients won't be switching back and forth on the ordering of the params that they request through agent since these are mainly going to be scripts or application services making the request.

Reply:
The problem with that argument is that you assume JSON marshaling behavior is consistent from language to language or library to library. It may be, but it may not. Go lexically orders JSON keys, but there's a performance impact to doing so, so speed demon libs or implementations may simply randomly order the parameters.

As an example, even in Go such applications may use jsoniter instead of stdlib json, and in jsoniter sorting the keys is an option that is not turned on by default.